OP-TEE Benchmark

igoropaniuk opened the pull request:
Related PRs:

Benchmark app:
https://github.com/linaro-swg/optee_benchmark/pull/1

Changes in OP-TEE
https://github.com/linaro-swg/linux/pull/38
https://github.com/OP-TEE/build/pull/124
https://github.com/OP-TEE/optee_os/pull/1365
https://github.com/OP-TEE/optee_client/pull/79

etienne-lms commented on core/arch/arm/pta/sta_benchmark.c:
```diff
@@ -0,0 +1,160 @@
+	DMSG("Unregistering benchmark, timestamp buffer paddr = %p\n",
+		(void *)virt_to_phys(bench_ts_buf));
+	bench_ts_buf = 0;
+	bench_stat_buf = 0;
```
minor: `NULL` rather than `0`. debug trace displays only the timestamp buffer, not the statistics one.

etienne-lms commented on core/arch/arm/pta/sta_benchmark.c:
```diff
@@ -0,0 +1,160 @@
+static void close_session(void *psess __unused)
+{
+
+}
```
minor: as create_ta/destroy_ta/open_session/close_session are stubs, you can simply discard them.

etienne-lms commented on core/arch/arm/pta/sta_benchmark.c:
```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright (c) 2015, Linaro Limited
```
minor: 2017

could you rename **sta_benchmark.c** as **pta_benchmarck.c**.  "static TA" is a deprecated reference.

etienne-lms commented on core/include/tee_bench_core.h:
```diff
@@ -0,0 +1,109 @@
+ */
+
+#ifndef TEE_BENCH_H
+#define TEE_BENCH_H
```
rather use `TEE_BENCH_CORE_H`, following the header file name.

etienne-lms commented on core/arch/arm/pta/sta_benchmark.c:
```diff
@@ -0,0 +1,160 @@
+#define BENCHMARK_CMD(id)			(0xFA190000 + (id & 0xFFFF))
+#define BENCHMARK_CMD_REGISTER_MEMREF	BENCHMARK_CMD(1)
+#define BENCHMARK_CMD_GET_MEMREF	BENCHMARK_CMD(2)
+#define BENCHMARK_CMD_UNREGISTER	BENCHMARK_CMD(3)
```
already defined in tee_bench_core.h

igoropaniuk replied:
right, will fix

etienne-lms commented on core/arch/arm/tee/entry_std.c:
```diff
@@ -290,13 +293,16 @@ static void entry_invoke_command(struct thread_smc_args *smc_args,
 out:
 	arg->ret = res;
 	arg->ret_origin = err_orig;
+
```
minor: useless change

etienne-lms commented on core/include/tee_bench_core.h:
```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2016, Linaro Limited
+ *
+ * This software is licensed under the terms of the GNU General Public
```
release this under the 2 clause BSD-like license.
you are still allowed to release it under a GPL like license in the scope of a linux driver/application.

etienne-lms commented on core/include/tee_bench_core.h:
```diff
@@ -0,0 +1,109 @@
+	asm volatile("adr %0, ." : "=r"(pc));
+#else
+	asm volatile("mov %0, r15" : "=r"(pc));
+#endif
```
i think `asm volatile("adr %0, ." : "=r"(pc));` works for both AArch32 and AArch64.

etienne-lms commented on core/include/tee_bench_core.h:
```diff
@@ -0,0 +1,109 @@
+static inline uint64_t read_ccounter(void)
+{
+	uint64_t ccounter = 0;
+#ifdef __aarch64__
```
minor: add empty line

etienne-lms commented on core/include/tee_bench_core.h:
```diff
@@ -0,0 +1,109 @@
+static inline __attribute__((always_inline)) uintptr_t read_pc(void)
+{
+	uintptr_t pc = 0;
+#ifdef __aarch64__
```
minor: add empty line

etienne-lms commented on core/include/tee_bench_core.h:
```diff
@@ -0,0 +1,109 @@
+	struct tee_time_buf *timeb = NULL;
+	uint64_t ts_i;
+
+	if ((cmd_id & BENCHMARK_CMD(0)) == BENCHMARK_CMD(0))
```
What about TAs that want to use such command IDs ? they won't be able to get benchmark stats.

etienne-lms commented on core/include/tee_bench_core.h:
```diff
@@ -0,0 +1,109 @@
+static inline __attribute__((always_inline))
+void bm_timestamp(uint32_t source, uint32_t cmd_id)
+{
+	struct tee_time_buf *timeb = NULL;
```
minor: useless init

etienne-lms commented on core/include/tee_bench_core.h:
```diff
@@ -0,0 +1,109 @@
+static void bm_timestamp(uint32_t source __unused,
+				uint32_t cmd_id __unused)
+{
+	;
```
add `inline`

minor: need this `;` ?

jenswi-linaro commented on lib/libutee/include/tee_api_defines.h:
```diff
@@ -450,8 +450,7 @@
 #define TEE_PARAM_TYPES(t0,t1,t2,t3) \
-   ((t0) | ((t1) << 4) | ((t2) << 8) | ((t3) << 12))
-
+		((t0) | ((t1) << 4) | ((t2) << 8) | ((t3) << 12))
```
Change not needed

jenswi-linaro commented on core/include/tee_bench_core.h:
```diff
@@ -0,0 +1,109 @@
+
+	if (timeb->tm_ind >= TEE_BENCH_MAX_STAMPS)
+		return;
+	ts_i = timeb->tm_ind++;
```
`timeb->tm_ind` should be copied before checking the value as normal world may have a window where it can update it unchecked.

What about concurrent updates?

jenswi-linaro commented on core/include/tee_bench_core.h:
```diff
@@ -0,0 +1,109 @@
+	if (!bench_ts_buf)
+		return;
+
+	timeb = (struct tee_time_buf *) bench_ts_buf;
```
Cast not needed

jenswi-linaro commented on core/include/tee_bench_core.h:
```diff
@@ -0,0 +1,109 @@
+}
+
+/* Adding timestamp to buffer */
+static inline __attribute__((always_inline))
```
With `__builtin_return_address(0)` instead below, or some clever macro to supply the return address as an argument to this function it wouldn't need to be inlined.

igoropaniuk commented on core/arch/arm/pta/sta_benchmark.c:
```diff
@@ -0,0 +1,160 @@
+		uint32_t nCommandID, uint32_t nParamTypes,
+		TEE_Param pParams[TEE_NUM_PARAMS])
+{
+	DMSG("command entry point for static ta \"%s\"", TA_NAME);
```
Not useful at all, just forgot to delete it :)

etienne-lms commented on core/arch/arm/pta/sta_benchmark.c:
```diff
@@ -0,0 +1,160 @@
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
```
x

etienne-lms replied:
xxx

jenswi-linaro commented on core/arch/arm/pta/benchmark.c:
```diff
@@ -0,0 +1,200 @@
+{
+	uint64_t ccounter = 0;
+#ifdef __aarch64__
+	asm volatile("mrs %0, PMCCNTR_EL0" : "=r"(ccounter));
```
Please add a `read_pmccntr_el0()` or similar in `arm64.h` and use that instead, and something similar for the ARM32 version below.

igoropaniuk replied:
Already moving it there :)

etienne-lms commented on core/arch/arm/pta/benchmark.c:
```diff
@@ -0,0 +1,200 @@
+	DMSG("Registering timestamp buffer, addr = %p, paddr = %" PRIxPA "\n",
+			p[0].memref.buffer,
+			virt_to_phys(p[0].memref.buffer));
+	bench_ts_global = p[0].memref.buffer;
```
Should check `p[0].memref.buffer` is mapped non-secure (case secure client) ?

etienne-lms replied:
If the client is secure, the buffer virtual address is likely to be temporary (to this TA invocation). Core should get the paddr and gets it core mapping to get a reliable virtual address for the buffer.

igoropaniuk replied:
@etienne-lms II didn't get it. Could you please provide any example of secure client case?
Or you mean TA-to-TA calls?

etienne-lms replied:
If a TA invokes your pTA, the memref parameters can be SHM references or secure memory reference (specific case when TA use a memref parameter on a TA private buffer).
As we expect here a SHM buffer (later RPC to tee-supplicant), maybe you can only check `tee_pbuf_is_non_sec(p[0].memref.buffer, p[0].memref.size)`.

igoropaniuk replied:
added check

etienne-lms commented on core/arch/arm/pta/benchmark.c:
```diff
@@ -0,0 +1,200 @@
+
+	/* fill timestamp data */
+	if (cur_cpu >= bench_ts_global->cores)
+		return;
```
+	thread_set_foreign_intr(true);

igoropaniuk replied:
See line https://github.com/OP-TEE/optee_os/pull/1365/files/7cc94f09527479a1bef1fca7f18f6fa6857bc2c7#diff-3a62a46da6ce718086e0bbc264a9f2f8R199

igoropaniuk replied:
or I can move call of `read_ccounter();` a bit higher (to `cur_cpu = get_core_pos();` ) to re-enable interrupt handling as soon as possible

igoropaniuk replied:
@etienne-lms

etienne-lms replied:
```c
	if (cur_cpu >= bench_ts_global->cores) {
		thread_set_foreign_intr(true); 
		return;
	}
```

igoropaniuk replied:
Ah, got it

etienne-lms commented on core/arch/arm/pta/benchmark.c:
```diff
@@ -0,0 +1,200 @@
+	cpu_buf = &bench_ts_global->cpu_buf[cur_cpu];
+	ts_i = __sync_fetch_and_add(&cpu_buf->head, 1);
+	ts_data.cnt = read_ccounter();
+	ts_data.addr = (uintptr_t) ret_addr;
```
minor: remove space before `ret_addr`

etienne-lms commented on core/arch/arm/tee/entry_std.c:
```diff
@@ -287,6 +288,8 @@ static void entry_invoke_command(struct thread_smc_args *smc_args,
 	struct tee_ta_session *s;
 	struct tee_ta_param param;
 
+	bm_timestamp();
```
This should be a known ID on which REE gets comprehensive.
I think this should be like `bm_timestamp(CORE_PRE/POST_IMVOKE_TA);`
Maybe this can come later... Specified bm points and free bm points.

igoropaniuk replied:
`__builtin_return_address(0);` in `bm_timestamp()` isn't sufficient?

igoropaniuk replied:
@etienne-lms

etienne-lms replied:
if the PC gives a clear view of trace point when rendering the output data to a human readable format, then it is fine.

igoropaniuk replied:
it's going to be mapped to specific file:line and function (the same as `addr2line` does)

jbech-linaro replied:
I think this is OK as it is for now.

etienne-lms commented on core/arch/arm/pta/benchmark.c:
```diff
@@ -0,0 +1,201 @@
+	if (!bench_ts_global)
+		return;
+
+	thread_set_foreign_intr(false);
```
What if called with foreign interruptions not enabled? If will be enabled at routine exit.

igoropaniuk replied:
good question, my first intention was to use `thread_restore_foreign_intr()`, but, actually, I didn't fully get where and how the previous CPSR value is saved (based on only one example of `thread_restore_foreign_intr()` I found in `void tee_svc_handler(struct thread_svc_regs *regs)` implementation), so the only one option was just to do it manually with `thread_get_exceptions`/`thread_set_exceptions`.

etienne-lms replied:
`uint32_t exceptions = thread_mask_exceptions(THREAD_EXCP_ALL); `
...
`thread_unmask_exceptions(exceptions); `

Or at least, mask `THREAD_EXCP_FOREIGN_INTR` as it is the noe that can makes thread to migrate to another cpu.

igoropaniuk replied:
thanks!

igoropaniuk commented:
@etienne-lms @jenswi-linaro Could check please also https://github.com/linaro-swg/optee_benchmark/pull/1/files, it includes almost 40-50% of the functionality. Thanks

jbech-linaro commented on core/arch/arm/pta/benchmark.c:
```diff
@@ -0,0 +1,199 @@
+
+	mutex_lock(&bench_reg_mu);
+
+	/* we accept only non-secure buffers, as we later perform
```
Nitpick: But I think checkpatch will complain here (multiline comment shall start with `/*` on a single line).

jbech-linaro commented on core/arch/arm/pta/benchmark.c:
```diff
@@ -0,0 +1,199 @@
+
+	/* we accept only non-secure buffers, as we later perform
+	 * registration of this buffer in NS layers
+	 * (optee linx kmod/optee client)
```
s/linx/Linux/

jbech-linaro commented on core/arch/arm/pta/benchmark.c:
```diff
@@ -0,0 +1,199 @@
+	exceptions = thread_mask_exceptions(THREAD_EXCP_ALL);
+	cur_cpu = get_core_pos();
+
+	/* fill timestamp data */
```
This comment seems a bit malplaced.

jbech-linaro commented on core/include/bench.h:
```diff
@@ -0,0 +1,84 @@
+#define TEE_BENCH_DIVIDER		64
+
+/* max amount of timestamps per buffer */
+#define TEE_BENCH_MAX_STAMPS	32
```
Isn't this a rather low default value? Does it mean what it says? I.e, there is room for 32 unique timestamps, then it will wrap?

igoropaniuk replied:
> I.e, there is room for 32 unique timestamps, then it will wrap?
right

jbech-linaro commented on core/include/bench.h:
```diff
@@ -0,0 +1,84 @@
+#define OPTEE_MSG_RPC_CMD_BENCH_REG_NEW		0
+#define OPTEE_MSG_RPC_CMD_BENCH_REG_DEL		1
+
+/* OP-TEE susbsystems ids */
```
Now I've started look into the PR in optee_os, but if this isn't a shared h-file (which I suspect it is), then we should maybe consider remove the ones not in use in optee_os. I'll update this comment when/if I've reviewed the other PRs.

igoropaniuk replied:
Each rep(besides `optee_benchmark`, it uses optee_client's 'public/tee_bench.h` file) has it's own version of bench.h (and it's not just a copy-paste), with our defines which are related only to that specific subsystem.
Or maybe I just didn't get your comment

jbech-linaro commented on core/include/bench.h:
```diff
@@ -0,0 +1,84 @@
+#ifdef CFG_TEE_BENCHMARK
+void bm_timestamp(void);
+#else
+/* stub */
```
It's already rather obvious that this is a stub.

jbech-linaro commented on core/include/optee_msg.h:
```diff
@@ -419,4 +419,5 @@ struct optee_msg_arg {
  */
 #define OPTEE_MSG_RPC_CMD_SHM_FREE	7
 
+#define OPTEE_MSG_RPC_CMD_BENCH_REG	20
```
I think this command also deserves a comment just likethe other cmd's here.

igoropaniuk replied:
will fix

jbech-linaro commented:
It'd be nice to have "some" documentation in a markdown file or similar, giving a quick high level overview of this. In the acceptance criteria (Jira), we have:
```
At optee_os/documentation there should be a file "benchmark.md". That describes, how to run the benchmark, how to interpret the data
```
That doesn't necessarily have to go into this PR, it could be done on a separate PR also.

etienne-lms commented on core/arch/arm/pta/benchmark.c:
```diff
@@ -0,0 +1,199 @@
+
+	/* fill timestamp data */
+	if (cur_cpu >= bench_ts_global->cores) {
+		thread_unmask_exceptions(exceptions);
```
This could panic the TEE. I guess we don't expect benchmark support in end product, so panicking here is not a issue.

etienne-lms commented on core/include/bench.h:
```diff
@@ -0,0 +1,84 @@
+#define BENCHMARK_CMD(id)	(0xFA190000 | ((id) & 0xFFFF))
+#define BENCHMARK_CMD_REGISTER_MEMREF	BENCHMARK_CMD(1)
+#define BENCHMARK_CMD_GET_MEMREF		BENCHMARK_CMD(2)
+#define BENCHMARK_CMD_UNREGISTER		BENCHMARK_CMD(3)
```
The 'pure' benchmark pTA API could go in a specific header file exported to TA devkit for being shared.
(I think Joakim had a comment about it)

igoropaniuk replied:
@etienne-lms Timestamping within TAs isn't supported yet, do we really need to move this header for now?

etienne-lms replied:
This is minor. Yet it is a good habit to locate pTA info definitions in the devkit so it can be used even by non-secure applications that already gets pTA info from the devkit, from **include/pta_xxx.h**, basically the pTA UUID and its command IDs.
See comment on `BENCHMARK_UUID` above.

etienne-lms commented on core/include/bench.h:
```diff
@@ -0,0 +1,84 @@
+#define BENCHMARK_CMD_GET_MEMREF		BENCHMARK_CMD(2)
+#define BENCHMARK_CMD_UNREGISTER		BENCHMARK_CMD(3)
+
+#define TEE_BENCH_DIVIDER		64
```
Maybe add a comment about what this macro is about.

igoropaniuk commented on core/arch/arm/pta/benchmark.c:
```diff
@@ -0,0 +1,199 @@
+	ret_addr = __builtin_return_address(0);
+
+	cpu_buf = &bench_ts_global->cpu_buf[cur_cpu];
+	ts_i = __sync_fetch_and_add(&cpu_buf->head, 1);
```
> Foreign interruptions (actually all interrupts) are already masked, so you can be pretty sure you can atomically update head value with even a complex SW sequence.

Agree. This came from the version, when there was only common timestamp buffer for all CPUs. I'll leave `__sync_fetch_and_add` only for NW EL0 apps (optee_client).

> Also, if head is lower than tail an reaches tail when incremented, it should update tail and maybe raise an overrun flag.

Right, we have already discussed it with Joakim, and he suggested to leave it for now as it is, and add support of overrun detection in further PRs. As for me, it's better just to add additioanl field to `struct tee_time_st` with will contain timestamp id (actually, it can be the `head` value, as it's incrementing all the time), so we can detect an over-run when parsing results by the actual timestamp analyzing app, which will detect a bit gap in timestamp IDs.

etienne-lms commented on core/arch/arm/include/arm64.h:
```diff
@@ -238,6 +238,16 @@ static __always_inline uint64_t read_fp(void)
+	return val;
+}
+
+
```
minor: empty line

etienne-lms commented on core/arch/arm/pta/benchmark.c:
```diff
@@ -0,0 +1,200 @@
+#include <string.h>
+#include <string_ext.h>
+#include <stdio.h>
+#include <trace.h>
```
minor: sort those includes.

etienne-lms commented on core/arch/arm/pta/benchmark.c:
```diff
@@ -0,0 +1,200 @@
+
+#define BENCHMARK_UUID \
+		{ 0x0b9a63b0, 0xb4c6, 0x4c85, \
+		{ 0xa2, 0x84, 0xa2, 0x28, 0xef, 0x54, 0x7b, 0x4e } }
```
minor: this could go in **lib/libutee/include/pta_benchmark.h** as other pTAs header files. 
See also comment on `BENCHMARK_CMD()` below.

etienne-lms commented on core/arch/arm/pta/benchmark.c:
```diff
@@ -0,0 +1,200 @@
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	mutex_lock(&bench_reg_mu);
```
lock the mutex after `if (!tee_pbuf_is_non_sec(...)) { ... }`

etienne-lms commented on core/arch/arm/pta/benchmark.c:
```diff
@@ -0,0 +1,200 @@
+	ret_addr = __builtin_return_address(0);
+
+	cpu_buf = &bench_ts_global->cpu_buf[cur_cpu];
+	ts_i = __sync_fetch_and_add(&cpu_buf->head, 1);
```
Is `__sync_fetch_and_add()` reliable ? A simple `cpu_buf->head++;` will fit.

etienne-lms commented on core/arch/arm/include/arm32.h:
```diff
@@ -601,6 +601,15 @@ static __always_inline uint32_t read_r7(void)
 	return val;
 }
+
+/* PMU cycle counter */
```
minor: useless comment

etienne-lms commented on core/arch/arm/include/arm64.h:
```diff
@@ -238,6 +238,15 @@ static __always_inline uint64_t read_fp(void)
 	return val;
 }
 
+/* PMU cycle counter */
```
minor: useless comment

etienne-lms commented on core/arch/arm/pta/benchmark.c:
```diff
@@ -0,0 +1,199 @@
+
+	res = rpc_reg_global_buf(OPTEE_MSG_RPC_CMD_BENCH_REG_DEL, 0, 0);
+
+	mutex_unlock(&bench_reg_mu);
```
can release lock before the RPC.

etienne-lms replied:
same comment as : useless (from core) but armless mutex hold. ok like this.

igoropaniuk replied:
@etienne-lms see https://github.com/OP-TEE/optee_os/pull/1365/files#r117725764

etienne-lms commented on core/arch/arm/pta/benchmark.c:
```diff
@@ -0,0 +1,199 @@
+	 * registration of this buffer in NS layers
+	 * (optee linux kmod/optee client)
+	 */
+	if (!tee_pbuf_is_non_sec(p[0].memref.buffer, p[0].memref.size))
```
`p[0].memref.buffer` is a virtual address. Should use `tee_vbuf_is_non_sec()`.

etienne-lms commented on core/arch/arm/pta/benchmark.c:
```diff
@@ -0,0 +1,199 @@
+			sizeof(struct tee_ts_cpu_buf) *
+			bench_ts_global->cores);
+
+	mutex_unlock(&bench_reg_mu);
```
i think the lock does not need to be held during the RPC.

igoropaniuk replied:
there is no any synchronization on that side (in the RPC handler in optee linux driver)

igoropaniuk replied:
@etienne-lms 
> there is no any synchronization on that side (in the RPC handler in optee linux driver)

etienne-lms replied:
As soon as `bench_ts_global` is stored, it is safe from core to use it, no ?
Anyway, holding the mutex a bit longer is not really armful. ok like this.

igoropaniuk replied:
I was trying to handle the case, when `register_benchmark_memref` is called twice (for the case, when someone has executed `optee_benchmark` app at the same time), so if don't to wrap `rpc_reg_global_buf` with mutex, it's possible to have a race condition in `static void handle_rpc_func_cmd_bm_reg(struct optee_msg_arg *arg)` [L312](https://github.com/linaro-swg/linux/pull/38/files#diff-9b23115d70507633758c1768dfe4de1bR312)

I think it's better to add condition in the beginning of `register_benchmark_memref` which will check if we have registered a buffer before, and then we can avoid such situation and unlock mutex before rpc call

igoropaniuk commented:
Rebased on latest master and tested

jforissier commented:
The commit description cannot be empty. Please explain what is meant by "benchmarking", how it is useful, how it is enabled (looks like it is always enabled? what if we want to reduce the size of the TEE? it is a debug feature after all, isn't it?), how one typically uses it, how it is implemented (high-level description, just to understand the interactions between the Secure/Non-secure kernel and user modes).

etienne-lms commented on core/arch/arm/pta/benchmark.c:
```diff
@@ -0,0 +1,199 @@
+		return;
+	}
+
+	/* fill timestamp data */
```
style: useless comment.

etienne-lms commented on core/arch/arm/tee/entry_std.c:
```diff
@@ -286,7 +287,10 @@ static void entry_invoke_command(struct thread_smc_args *smc_args,
+	bm_timestamp();
+
 	res = copy_in_params(arg->params, num_params, &param);
+
```
remove this empty line.

etienne-lms commented on core/include/bench.h:
```diff
@@ -0,0 +1,82 @@
+ */
+#define TEE_BENCH_DIVIDER		64
+
+/* max amount of timestamps per buffer */
```
style: start comment with a uppercase character.

igoropaniuk commented:
@jforissier 
> The commit description cannot be empty. Please explain what is meant by "benchmarking", how it is useful, how it is enabled (looks like it is always enabled? what if we want to reduce the size of the TEE? it is a debug feature after all, isn't it?), how one typically uses it, how it is implemented (high-level description, just to understand the interactions between the Secure/Non-secure kernel and user modes).

I'm going to add some basic information to this commit, that I'll create additional PR with extended Markdown doc with detailed information about its implementation, usage etc. Deal? :)

jforissier commented:
@igoropaniuk yep sounds good to me, thanks.

jforissier commented:
BTW @igoropaniuk if you want you can create and push an empty commit (git commit --allow-empty) so that we can review the description.

igoropaniuk commented:
@jforissier I've already amended current commit, and added description there (I hope it can be called a "description" :) )

igoropaniuk commented:
@etienne-lms @jforissier 
Any updates?

igoropaniuk commented:
@etienne-lms all comments addressed

igoropaniuk commented:
I think I've addressed all comments so far and now think this is ready for a final review and if there are no more comments we should consider merging this now

igoropaniuk commented:
@etienne-lms @jbech-linaro @jenswi-linaro @jforissier 
Thanks for all your comments. Is there anything else worth mentioning/fixing?

jforissier commented:
Abbreviating "benchmark" as "bench" sounds weird I think (bench.h, TEE_BENCH_* etc.) and is not done consistently, so it would be better to write "benchmark" always.
Anyway:
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

etienne-lms commented:
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

igoropaniuk commented:
Applied tags and rebased

igoropaniuk commented:
@jforissier all these PRs are rebased on master, tags applied and ready for merge

https://github.com/linaro-swg/optee_benchmark/pull/1
https://github.com/linaro-swg/linux/pull/38
https://github.com/OP-TEE/build/pull/124
https://github.com/OP-TEE/optee_os/pull/1365
https://github.com/OP-TEE/optee_client/pull/79

thanks!

jforissier commented:
All merged, thanks.

jbech-linaro commented:
Thanks all reviewers and big thanks to you @igoropaniuk for implementing this and having the patience of reworking it a couple of times.
